src/Microsoft.AspNet.Razor.Runtime/BufferEntryCollection.cs

```diff
@@ -0,0 +1,191 @@
+        public void Add(string value)
+        {
+            _buffer.Add(value);
+            _isModified = true;
+        }
+
+        /// <summary>
+        /// Adds a subarray of characters to the buffer.
+        /// </summary>
+        /// <param name="value">The array to add.</param>
+        /// <param name="index">The character position in the array at which to start copying data.</param>
+        /// <param name="count">The number of characters to copy.</param>
+        public void Add([NotNull] char[] value, int index, int count)
+        {
+            if (index < 0)
+            {
+                throw new ArgumentOutOfRangeException("index");
+            }
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException("count");
+            }
+            if (value.Length - index < count)
+            {
+                throw new ArgumentOutOfRangeException("count");
+            }
+
+            while (count > 0)
+            {
+                // Split large char arrays into 1KB strings.
+                var currentCount = Math.Min(count, MaxCharToStringLength);
+                Add(new string(value, index, currentCount));
+                index += currentCount;
+                count -= currentCount;
+            }
+            _isModified = true;
+        }
+
+        /// <summary>
+        /// Adds an instance of <see cref="BufferEntryCollection"/> to the buffer.
+        /// </summary>
+        /// <param name="buffer">The buffer collection to add.</param>
+        public void Add([NotNull] BufferEntryCollection buffer)
+        {
+            _buffer.Add(buffer.BufferEntries);
+            _isModified = true;
+        }
+
+        public void Clear()
+        {
+            _buffer.Clear();
+            _isModified = true;
+        }
+
+        /// <inheritdoc />
+        public IEnumerator<string> GetEnumerator()
+        {
+            return new BufferEntryEnumerator(_buffer);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        private sealed class BufferEntryEnumerator : IEnumerator<string>
+        {
+            private readonly Stack<IEnumerator<object>> _enumerators = new Stack<IEnumerator<object>>();
+            private readonly List<object> _initialBuffer;
+
+            public BufferEntryEnumerator(List<object> buffer)
+            {
+                _initialBuffer = buffer;
+                Reset();
+            }
+
+            public IEnumerator<object> CurrentEnumerator
+            {
+                get
+                {
+                    return _enumerators.Peek();
+                }
+            }
+
+            public string Current
+            {
+                get
+                {
+                    var currentEnumerator = CurrentEnumerator;
+                    Debug.Assert(currentEnumerator != null);
+
+                    return (string)currentEnumerator.Current;
```

Comment:
I see that this existed in the old version of the `BufferEntryCollection`, I thought I saw why it didn't explode but then realized was wrong again.

```C#
var foo = new BufferEntryCollection();
var bar = new BufferEntryCollection();
bar.Add(foo);
var enumerator = bar.GetEnumerator();
var current = enumerator.Current; // I think this blows up
```